implement list codec

koubaa opened the pull request:
### What does this implement/fix? Explain your changes.
Implement conversions from python lists, iterables, and sequences to appropriate C# interfaces
...

### Does this close any currently open issues?
Probably a few of them. I'll look for it later.


### Any other comments?
I've done the codec for three levels of collections.  roughly I map these:
| python | C# |
|---|---|
| iterable | IEnumerable |
| sequence | ICollection |
| list | IList |

The mapping isn't exactly perfect, naturally.  But this is the best I think I can do right now.

I need to add some more codec unit tests and functional embedding tests.

### Checklist

Check all those that are applicable and complete.

-   [ ] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [x] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [ ] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

koubaa commented:
@amos402 @filmor @lostmsu please review.  I think this PR may help with our implicit conversions of lists.

lostmsu commented:
Haven't done a full pass yet, but instead of doing "levels" you can just have 3 decoders, and register them in order from highest priority to lowest priority. E.g.
```csharp
Register(ListEncoder.Instance);
Register(CollectionDecoder.Instance);
Register(IterableDecoder.Instance);
```

koubaa commented:
@lostmsu I can change the physical layout but wanted to make sure I was on the right track first.  I like the idea of having multiple codecs with priority, that way clients can choose which encoder they need.  I'll work on that.

koubaa commented on src/embed_tests/Codecs.cs:
```diff
@@ -82,5 +82,130 @@ public class Codecs {
+
+            //we'd have to copy into a list to do this.  not the best idea to support it.
+            //maybe there can be a flag on listcodec to allow it.
+            Assert.IsFalse(codec.CanDecode(x, typeof(List<int>)));
```
@lostmsu @filmor With your idea of multiple codecs I can have a DeepCopyListCodec that takes any ienumerable and converts it by copying to a list.  This is what we do right now in master and if we get rid of that implicit conversion the user can just register this codec if that causes an undesirable behavior change for them.

lostmsu replied:
I am thinking, that `Python.Runtime.dll` should only contain codecs, that perform lossless conversions, perhaps even by default. E.g. if a type is exactly `builtins.list` it is safe to convert it to `IList<T>` via some safe wrapper type `PythonList<T>`, that would contain a reference to the original object. Preferably, when this wrapper is passed back to Python it should either just pass the wrapped object, or behave exactly like a normal `list` would.

The codecs, that are not lossless, like converting a `builtins.list` to `T[]`, and codecs, that act on relatively uncommon types like `System.Uri` should go to a separate extra package (created yesterday after seeing your PR): https://github.com/pythonnet/codecs

koubaa replied:
@lostmsu done.  I didn't go as far as to enable by default but that can be done as a followup commit.  Please look closely at the test cases as I tried to capture the behavior of each of the three codecs to make sure it meets your expectation.

koubaa commented:
This is related to #912

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,52 @@
+        public IterableWrapper(PyObject pyObj)
+        {
+            pyObject = pyObj;
+            iterObject = new PyObject(Runtime.PyObject_GetIter(pyObj.Handle));
```
`iterObject` should be local to `GetEnumerator` implementations, otherwise you won't be able to enumerate this instance twice.

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+        {
+            get
+            {
+                IntPtr item = Runtime.PySequence_GetItem(pyObject.Handle, index);
```
I wonder if there would be any difference between using `PySequence_GetItem` vs `PyList_GetItem` vs `PyObject_GetItem` here.

koubaa replied:
@lostmsu good point.  I should use PyList_GetItem in ListWrapper and PySequence_GetItem in SequenceWrapper.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,52 @@
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            IntPtr item;
+            while ((item = Runtime.PyIter_Next(iterObject.Handle)) != IntPtr.Zero)
```
If `PyIter_Next` fails, this should check the reason. If it is not `StopIteration`, an exception should be passed to .NET

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+                var result = Runtime.PySequence_SetItem(pyObject.Handle, index, pyItem);
+                Runtime.XDecref(pyItem);
+                if (result == -1)
+                    throw new Exception("failed to set item");
```
You should raise the actual Python exception.

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+
+            IntPtr pyItem = Converter.ToPython(item, typeof(T));
+            if (pyItem == IntPtr.Zero)
+                throw new Exception("failed to insert item");
```
Same, raise the actual Python exception

koubaa replied:
@lostmsu Converter.ToPython registers a python exception if it fails?

lostmsu replied:
@koubaa you'd have to take a look at it.

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+        public void Insert(int index, T item)
+        {
+            if (IsReadOnly)
+                throw new NotImplementedException();
```
I think this should be `InvalidOperationException("Collection is read-only")`

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+            var result = Runtime.PyList_Insert(pyObject.Reference, index, pyItem);
+            Runtime.XDecref(pyItem);
+            if (result == -1)
+                throw new Exception("failed to insert item");
```
Python exception

lostmsu commented on src/runtime/CollectionWrappers/SequenceWrapper.cs:
```diff
@@ -0,0 +1,98 @@
+        {
+            for (int index = 0; index < Count; index++)
+            {
+                array[index + arrayIndex] = getItem(index);
```
Not sure, but it might be more effective to use iterator.

koubaa replied:
@lostmsu not disagreeing but please define effective.  Code clarity or runtime performance?

lostmsu replied:
@koubaa , I meant runtime performance. I am not sure if Python sequences must provide O(1) indexing.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,52 @@
+
+        public IterableWrapper(PyObject pyObj)
+        {
+            pyObject = pyObj;
```
I recommend check the argument for `null`

lostmsu replied:
Why is this resolved? Is there a legitimate scenario for `IterableWrapper` wrapping `None`/`null`?

koubaa replied:
@lostmsu This is how I interpreted your comment.  did you mean to say I should throw on a null input?

lostmsu replied:
@koubaa this is right. Though you should throw `ArgumentNullException`

lostmsu commented on src/runtime/CollectionWrappers/SequenceWrapper.cs:
```diff
@@ -0,0 +1,98 @@
+        {
+            get
+            {
+                return (int)Runtime.PySequence_Size(pyObject.Handle);
```
`PySequence_Size` might fail, according to documentation.

koubaa replied:
@lostmsu true, but it doesn't say it throws, so I will throw a c# exception in this case

lostmsu commented on src/runtime/CollectionWrappers/SequenceWrapper.cs:
```diff
@@ -0,0 +1,98 @@
+                throw new NotImplementedException();
+            var result = Runtime.PySequence_DelSlice(pyObject.Handle, 0, Count);
+            if (result == -1)
+                throw new Exception("failed to clear sequence");
```
Python exception

koubaa commented:
@lostmsu While the unit tests work when I tried it from python I realized (again) that CanDecode is based on the type rather than the instance, so I changed that.  Also I added some python tests

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,79 @@
+        private void propagateIterationException()
+        {
+            var err = Runtime.PyErr_Occurred();
+            if (err != null && err != Exceptions.StopIteration)
```
Documentation recommends using PyErr_ExceptionMatches

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,79 @@
+            pyObject = pyObj;
+        }
+
+        private void propagateIterationException()
```
I think this will keep `StopIteration` "raised". E.g. next check for unhandled Python exception will return it.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,79 @@
+{
+    internal class IterableWrapper<T> : IEnumerable<T>
+    {
+        protected PyObject pyObject;
```
This can be made `readonly`. Then you can drop null checks in `GetEnumerator`

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,79 @@
+        public IEnumerator<T> GetEnumerator()
+        {
+            if (pyObject == null) yield break;
+            PyObject iterObject = null;
```
Ideally, you should dispose this object.

koubaa replied:
@lostmsu shouldn't its finalizer do that?

lostmsu replied:
When we know the lifetime, we should call `Dispose` at the end of it to reduce the load on GC

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,63 @@
+            get
+            {
+                var item = Runtime.PyList_GetItem(pyObject.Handle, index);
+                var pyItem = new PyObject(item);
```
Dispose this

koubaa replied:
@lostmsu again I thought the point of making it IDisposable was that the finalizer would take care of that

amos402 replied:
Wrap it by using statement is better. If you knew it's no use anymore you should dispose it manually instead of waiting for the auto collection to do it.

koubaa replied:
@lostmsu @amos402 sorry, I might be missing context (I rebased and I don't know if this comment lines up with the code it was pointing to initially).

What specifically should I be disposing or wrap in a using statement?

amos402 replied:
```c#
using (var pyItem = new PyObject(item)) { }
```

lostmsu replied:
@koubaa this still needs a fix. With C# 8+ which we use now, it should just be `using var pyItem = new PyObject(item);` so that the Python object, corresponding to the `item` would be disposed and not pollute finalizer thread unless duplicated.

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,63 @@
+            {
+                IntPtr pyItem = Converter.ToPython(value, typeof(T));
+                if (pyItem == IntPtr.Zero)
+                    throw new Exception("failed to set item");
```
Perhaps `InvalidCastException`. Also, might need to supply the cause (e.g. underlying `PythonException`).

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,63 @@
+                if (pyItem == IntPtr.Zero)
+                    throw new Exception("failed to set item");
+
+                var result = Runtime.PyList_SetItem(pyObject.Handle, index, pyItem);
+                Runtime.XDecref(pyItem);
```
According to docs https://docs.python.org/3.5/c-api/list.html#c.PyList_SetItem decrefs internally.

lostmsu commented on src/runtime/CollectionWrappers/SequenceWrapper.cs:
```diff
@@ -0,0 +1,108 @@
+                if (size == -1)
+                {
+                    Runtime.CheckExceptionOccurred();
+                    throw new Exception("Unable to get sequence size!");
```
`CheckExceptionOccured` should throw in the line above

koubaa replied:
@lostmsu according to python docs for PySequenceDelSlice:
Delete the slice in sequence object o from i1 to i2. Returns -1 on failure.

It doesn't explicitly say it will set an error on failure.  Is this implicit anywhere I see the word failure in the python doc?

lostmsu commented on src/runtime/CollectionWrappers/SequenceWrapper.cs:
```diff
@@ -0,0 +1,108 @@
+            if (index >= Count || index < 0)
+                throw new IndexOutOfRangeException();
+
+            return Runtime.PySequence_DelItem(pyObject.Handle, index) != 0;
```
Failure of `PySequence_DelItem` is an exception too.

koubaa replied:
@lostmsu in that case I will clear the exception in Remove (In C# it is idiomatic to return false in Remove rather than to throw) but throw it in ListWrapper RemoveAt

lostmsu replied:
That depends on the reason for the failure. It should not throw only if the item was not in the collection in the first place. If it was in the collection, but something prevented it from being removed, it must throw.

lostmsu commented on src/runtime/converterextensions.cs:
```diff
@@ -156,7 +156,7 @@ bool TryDecode(IntPtr pyHandle, out object result)
         #endregion
 
-        internal static void Reset()
+        public static void Reset()
```
Is there a scenario when this should be done outside of runtime shutdown?

koubaa replied:
@lostmsu Doubt it.  I just needed it to clear the state between unit test runs.

lostmsu replied:
Yes, I think you should be shutting down the runtime.

amos402 commented:
If you're considering the performance improvement, this branch may help you.
https://github.com/joonhwan/pythonnet/tree/fast-call
It's an ancient branch I deleted a long time ago, my current internal branch has far more than that.
You can do some performance tests, like method calls, properties get-set, etc, you will see that the branch above has a lot faster than the master now. The simple points are avoiding any `MethodInfo.Invoke`, use pre-created delegates instead, prevent any boxing or unboxing if you can. After that, if you want a higher than it, code generation is an option.

The `codec` has some design points can improve:
* `Decode`, `Encode` used `object`, which means it may occur boxing.
* When it `decode/encode` an object, it has iteration operations which can be avoided, or use some hash table for getting the coder directly.
* For achieving no reflection calling, no boxing or unboxing, you may need to redesign the method calling mechanism just like I did, you can check the `methodobject.cs` on the above branch.

koubaa reviewed (commented):
@lostmsu could you please take another look at this?

koubaa commented:
> If you're considering the performance improvement, this branch may help you.
> https://github.com/joonhwan/pythonnet/tree/fast-call
> It's an ancient branch I deleted a long time ago, my current internal branch has far more than that.
> You can do some performance tests, like method calls, properties get-set, etc, you will see that the branch above has a lot faster than the master now. The simple points are avoiding any `MethodInfo.Invoke`, use pre-created delegates instead, prevent any boxing or unboxing if you can. After that, if you want a higher than it, code generation is an option.
> 
> The `codec` has some design points can improve:
> 
> * `Decode`, `Encode` used `object`, which means it may occur boxing.
> * When it `decode/encode` an object, it has iteration operations which can be avoided, or use some hash table for getting the coder directly.
> * For achieving no reflection calling, no boxing or unboxing, you may need to redesign the method calling mechanism just like I did, you can check the `methodobject.cs` on the above branch.

Thank you for sharing!  I haven't measured performance issues but now I know what to try if I do.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,95 @@
+        private void propagateIterationException()
+        {
+            var err = Runtime.PyErr_Occurred();
+            if (err == null) return;
```
`PyErr_Occurred` returns `IntPtr`. You can't compare it to `null`, you have to compare it to `IntPtr.Zero`.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,95 @@
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            PyObject iterObject = null;
+            using (Py.GIL())
```
Actually, it is a job of the caller of this method to hold GIL for any call on `GetEnumerator` or `MoveNext`

koubaa replied:
@lostmsu I defer to you but I am not convinced.  If I am passing an instance of a python list using this wrapper (let's say as an IEnumerable<T>) to some 3rd party library which is going to iterate over it at some point in the future, I would have to worry about when that library actually does the enumeration and enclose that with holding the GIL.

Personally I think it makes more sense to consider that the contract of IEnumerable<T> is "An object which can be iterated".  I think adding holding the GIL to the expectation of whoever is given an IEnumerable<T> that happens to be an IterableWrapper<T> isn't idiomatic C#.

lostmsu replied:
> it at some point in the future, I would have to worry about when that library actually does the enumeration and enclose that with holding the GIL.

I think you're right here.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,95 @@
+            while (true)
+            {
+                IntPtr item = IntPtr.Zero;
+                using (Py.GIL())
+                {
```
Same: GIL has to be held by the caller.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,95 @@
+            var err = Runtime.PyErr_Occurred();
+            if (err == null) return;
+
+            //remove StopIteration exceptions
+            if (0 != Runtime.PyErr_ExceptionMatches(Exceptions.StopIteration))
+            {
+                Runtime.PyErr_Clear();
+                return;
+            }
```
According to the official documentation, `PyIter_Next` should not be setting `StopIteration` exception:
https://docs.python.org/3.5/c-api/iter.html#c.PyIter_Next
> If there are no remaining values, returns NULL with no exception set

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,95 @@
+            Runtime.CheckExceptionOccurred();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
```
Because `IEnumerator<T>` also implements `IEnumerator`, you don't actually need this method to have full body. You can just replace it with `IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();`

koubaa replied:
@lostmsu thanks.  I was trying to remember what the shorthand for this was when I was implementing it but couldn't and just duplicated it.  I knew it was something this simple..

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,71 @@
+                var item = Runtime.PyList_GetItem(pyObject.Reference, index);
+                var pyItem = new PyObject(item);
+
+                if (!Converter.ToManaged(pyItem.Handle, typeof(T), out object obj, true))
+                    Runtime.CheckExceptionOccurred();
+
+                return (T)obj;
```
This can be replaced with `return pyItem.As<T>();`

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,71 @@
+            }
+            set
+            {
+                IntPtr pyItem = Converter.ToPython(value, typeof(T));
+                if (pyItem == IntPtr.Zero)
+                {
+                    throw new InvalidCastException(
+                        "cannot cast " + value.ToString() + "to type: " + typeof(T).ToString(),
+                        new PythonException());
+                }
```
Why not just call `value.ToPython()` here?

lostmsu commented on src/runtime/CollectionWrappers/ListWrapper.cs:
```diff
@@ -0,0 +1,71 @@
+                throw new PythonException();
+
+            var result = Runtime.PyList_Insert(pyObject.Reference, index, pyItem);
+            Runtime.XDecref(pyItem);
```
Doing a decref with error set might have unintended consequences.

koubaa replied:
@lostmsu with value.ToPython() I don't have to manually mess with the ref count

koubaa commented:
@lostmsu would you mind reviewing this again?  I resolved all of your previous comments

lostmsu commented:
@koubaa sure, but can't do it until Saturday.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,42 @@
+                    var item = Runtime.PyIter_Next(iterObject.Handle);
+                    if (item == IntPtr.Zero)
+                    {
+                        iterObject.Dispose();
```
You must check if there was an exception here.

https://docs.python.org/3.8/c-api/iter.html#c.PyIter_Next

lostmsu commented on src/runtime/CollectionWrappers/SequenceWrapper.cs:
```diff
@@ -0,0 +1,101 @@
+
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            var index = 0;
```
Please, check inputs, and, if possible that the collection will fit before making an attempt to copy.

Generally for any operation it is highly preferred that it either completes successfully, or does not make any changes to its inputs in case of exception.

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Python.Runtime.CollectionWrappers
+{
+    internal class IterableWrapper<T> : IEnumerable<T>
```
To be a lossless conversion, when passed back to Python an instance of this class should pass the `pyObject` field. I think this behavior must be added along with decoders, that create it.

koubaa replied:
@lostmsu Right now I only defined decoders.  I would think this is the responsibility of the code handling object conversion back to python (i.e. encoders) - not the collection wrappers themselves.

In this branch these decoders are used by default, but in a future PR which replaces implicit copy list conversion with them, we can make sure to handle this.

koubaa commented:
@lostmsu I'll look to enable these by default (and handle issues associated w/that) in a later PR

koubaa commented:
@lostmsu I rebased to fix the merge conflict.  Is there anything else needed before this can be merged?

lostmsu commented on src/embed_tests/Codecs.cs:
```diff
@@ -57,31 +66,262 @@ public class Codecs {
+            var pyTuple = new PyObject(Runtime.PyTuple_New(3));
+            var pyTupleType = pyTuple.GetPythonType();
+
+            Runtime.PyTuple_SetItem(pyTuple.Handle, 0, items[0].Handle);
+            Runtime.PyTuple_SetItem(pyTuple.Handle, 1, items[1].Handle);
+            Runtime.PyTuple_SetItem(pyTuple.Handle, 2, items[2].Handle);
```
This will cause double-free with possible subsequent access violation in tests, as `PyTuple_SetItem` steals the item reference.

I suggest to use `pyTuple[i] = item` instead.

christabella commented:
@koubaa do you plan to finish this PR? The open issues that this would close include
#909 #601 #962

koubaa commented:
@christabella I have been waiting for @lostmsu to merge it

koubaa commented:
@lostmsu I rebased the changes to resolve the conflicts.  Is there anything needed before this can be merged?

lostmsu commented:
@koubaa @christabella once the CI tests pass, I can take this. Note, however, that I have a different approach in mind, based on [mixins and base class faking](https://github.com/losttech/pythonnet/tree/losttech/master/src/runtime/Mixins) (a feature I have in my fork, that is yet to be upstreamed, its implementation is relatively involved). So this will probably be replaced before 3.0 is released.

koubaa commented:
@lostmsu the CI is fixed

lostmsu commented on src/embed_tests/Codecs.cs:
```diff
@@ -57,31 +66,258 @@ public class Codecs {
             }
         }
+
+        static string GetIntIterableCommands(string instanceName)
+        {
+            var builder = new System.Text.StringBuilder();
+            builder.AppendLine(@"
+class foo():
+    def __init__(self):
+        self.counter = 0
+    def __iter__(self):
+        return self
+    def __next__(self):
+        if self.counter == 3:
+            raise StopIteration
+        self.counter = self.counter + 1
+        return self.counter");
+
+            builder.AppendLine(instanceName + " = foo()");
+            return builder.ToString();
+        }
+
+        static PyObject GetPythonIterable()
+        {
+            var locals = new PyDict();
+            using (Py.GIL())
+            {
+                PythonEngine.Exec(GetIntIterableCommands("foo_instance"), null, locals.Handle);
+            }
+
+            return locals.GetItem("foo_instance");
+        }
```
I think you can simply do `Eval("map(lambda x: x, [1,2,3])")`.

koubaa replied:
@lostmsu wow that is clever, nice!

lostmsu commented on src/runtime/Codecs/IterableDecoder.cs:
```diff
@@ -0,0 +1,55 @@
+
+        internal static bool IsIterable(PyObject objectType)
+        {
+            return objectType.HasAttr("__iter__");
```
Can this be replaced with a call to `PyIter_Check`?

koubaa replied:
Yes it can

koubaa replied:
@lostmsu actually as it turns out, no it can't.  It leads to some test failures.
https://github.com/pythonnet/pythonnet/pull/1084/checks?check_run_id=1930584148

The reason is that the codec only has the *type*, not the *instance*.  PyIter_Check is meant to be called on the instance.

lostmsu replied:
Right. OK then.

lostmsu commented on src/runtime/Codecs/ListDecoder.cs:
```diff
@@ -0,0 +1,49 @@
+            if (!SequenceDecoder.IsSequence(objectType)) return false;
+
+            //returns wheter the type is a list.
+            return objectType.Handle == Runtime.PyListType;
```
Seems like this check is the only one needed.

koubaa replied:
@lostmsu True, but eventually this should support any object that implements the list protocol (e.g. numpy arrays).

lostmsu commented on src/runtime/CollectionWrappers/IterableWrapper.cs:
```diff
@@ -0,0 +1,43 @@
+        {
+            if (pyObj == null)
+                throw new ArgumentNullException();
+            pyObject = pyObj;
```
I believe you should make a new `PyObject` here, pointing to the same object in Python (e.g. `pyObject = new PyObject(pyObj.Reference)`). The reason is that if the original `PyObject` is disposed (as suggested below in `ListWrapper.this[int index]`), this iterator will become useless.

koubaa replied:
@lostmsu thanks, I often forget that C# objects can be explicitly disposed before they get axed by the GC.
